BIP324: Enable v2 P2P encrypted transport

dhruv opened the pull request:
This PR brings together all other BIP324 PRs and enables v2 encrypted P2P transport.

Builds on top of PRs: #25361, #23233, #23561, #23432. It looks like there's a lot of commits, but **only the last 12 commits belong in this PR**. The rest will be merged with upstream PRs.

The dependency tree for BIP324 PRs is [here](https://bip324.com/sections/code-review/).

BIP324 is [here](https://github.com/bitcoin/bips/blob/master/bip-0324.mediawiki).

### Running a v2 node
#### Get the code
```
git remote add bip324 git@github.com:dhruv/bitcoin.git
git fetch bip324
git checkout bip324/bip324-enable
```

#### Build for your OS

Follow the appropriate instructions [here](https://github.com/bitcoin/bitcoin/blob/master/doc/README.md)

#### Run the node

```
src/bitcoind -conf=CONFIG_FILE -v2transport=1
```

#### Connect with a friend's v2 node
```
src/bitcoin-cli -conf=CONFIG_FILE addnode "FRIEND_IP:FRIEND_PORT" "add" true
```
The last parameter(`p2p_v2:true`) signals to your node that the peer is running a v2 supportive client and we should attempt to make an encrypted P2P connection (you're simulating the `NODE_P2P_V2` service flag advertisement manually). Should that fail however (say because the peer told you mistakenly, lied, etc.), this code will downgrade the connection to unencrypted v1 transport.

### Things you are helpful to test

* If your friend's node is a v2 node, you can see with wireshark that the bytes are pseudorandom (the easiest way to confirm this is that with a v1 connection, wireshark will tell you it has detected a Bitcoin connection and it'll even parse out the metadata like message type, etc; with v2, wireshark has no idea -- of course that _could_ be because wireshark does simply not know v2, but it is because the bytestream is pseudorandom)
* Compare the v2 encrypted session id exposed via `getpeerinfo` as `v2_session_id` with your friend.
* Add another peer that is actually v1, but try `addnode` still indicating v2 support. You should see with wireshark that after a failed attempt at a v2 handshake, the connection is downgraded to unencrypted v1 and wireshark can parse it.

I've been told there are v2 nodes running at (happy to update the list as more people run persistent v2 nodes; message me and I'll add it here):
```
be.anyone.eu.org
rp7k2go3s5lyj3fnj6zn62ktarlrsft2ohlsxkyd7v3e3idqyptvread.onion:8333
jdcoysubtxazi7dketpyb5rnjorvxad4onftveohash2pdwkgw4bvnqd.onion:8333
xci6cphki2pdb5qe7axzrcxcxabkbm24z4zlv2hn4ziy6grquqco2kyd.onion:8333
slvtesfgg3mkksqqzh67al4sq6dx3rhlzqepa4ny7jonzuckg6msf3id.onion:8333
gifm4fnj3vua664xhgeanx5fnpco3txkqy4amr4txbfsciiyrkxpf2qd.onion:8333
300:5ecb:6b8a:d837::3:8333
300:5ecb:6b8a:d837::a6d6:8333
2001:470:1f1a:365::2:8333
2001:470:1f1b:365:aa20:66ff:fe3f:1909:8333
184.74.240.157:8533
95.179.145.232:8333
```

dhruv committed:
Add BIP324 short-IDs to protocol.cpp

Co-authored-by: Jonas Schnelli <dev@jonasschnelli.ch>

dhruv committed:
Add BIP324 v2 transport serializer and deserializer

Co-authored-by: Jonas Schnelli <dev@jonasschnelli.ch>

dhruv committed:
BIP324 Cipher Suite

dhruv committed:
fuzz: Add fuzz test for v2 transport {de}serialization

dhruv committed:
Encode CKey to ElligatorSwift representation

dhruv committed:
Bench tests for CKey->EllSwift

dhruv committed:
Fuzz tests for CKey->EllSwift

dhruv committed:
HKDF key derivation from ECDH secret for BIP324

dhruv committed:
Fuzz test for BIP324 key derivation

dhruv committed:
p2p: Advertise v2 transport if CLI arg is on

dhruv committed:
rpc: addnode arg to use BIP324 v2 p2p

dhruv committed:
p2p: Use v2 transport between supportive peers

dhruv committed:
scripted-diff: rename use_v2 to use_addr_v2 in src/protocol.h

-BEGIN VERIFY SCRIPT-
sed -i 's/use_v2/use_addr_v2/g' ./src/protocol.h
-END VERIFY SCRIPT-

dhruv committed:
p2p: BIP324 transport version messages

dhruv committed:
refactor: Add InitP2P() to NetEventsInterface

dhruv committed:
p2p: BIP324 v2.0 clients retry with v1 protocol

dhruv commented:
Bringing in [upstream changes](https://github.com/bitcoin/bitcoin/pull/20962#issuecomment-1074367550) from #20962. Ready for further review.

dhruv commented:
Rebased to resolve conflicts with #21160.

Addressed comments from @mzumsande and @kallewoof in #23900.

Ready for further review.

dhruv commented:
Rebased. Ready for further review.

jonatack commented:
Concept ACK

laanwj commented:
I'm testing this on one of my nodes. Feel free to message me if you want to connect.

Is there any way to see (in, say, `getpeerinfo`) whether a connection uses P2P v2?

dhruv commented:
Thanks, @laanwj !

> Is there any way to see (in, say, getpeerinfo) whether a connection uses P2P v2?

That's a good idea, I've added it here: https://github.com/bitcoin/bitcoin/projects/17#card-80335229 I'll work on it once I'm done with the basics.

dhruv commented:
Rebased. `git range-diff e0680bbce f7cae59 d0aef7a1d`

Ready for further review.

uvhw commented:
Vào CN, 10 thg 4, 2022 lúc 2:47 SA DrahtBot ***@***.***> đã
viết:

> 🐙 This pull request conflicts with the target branch and needs rebase
> <https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#rebasing-changes>
> .
>
> Want to unsubscribe from rebase notifications on this pull request? Just
> convert this pull request to a "draft".
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/24545#issuecomment-1094114277>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AVQHDXC57RFNZXC7DCRHKVDVEHND7ANCNFSM5QSHGKNQ>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>
-- 
~~~~1GwvLW9qJ8uaYjew3cFvPiqxViWhuU1pKT~~BTC1505~~

0xB10C commented:
I'm running d0aef7a1d2f324aad923f734247d34baf9a50ae8 on the node behind https://bitcoind.observer. DM me for the IP.

0xB10C commented:
From what I can tell, my node fails to connect to another p2p v2 peer (likely @laanwj's, we haven't fully confirmed that yet) and stops accepting inbound messages from any other peer cutting it off from the network entirely.

dhruv commented:
@0xB10C Thanks for the logs. I'll look into it. We are making some large changes to the spec at the moment, so I'm working on updating the BIP and some of the branches are actually out of sync now. I'll be back in touch soon with updated code.

dhruv committed:
RFC8439 nonce and counter for ChaCha20

theStack commented:
@dhruv: I'd have interest in testing and reviewing BIP324. Is there anything useful we can do for now or is it better to just wait for the final spec / updated BIP? (Maybe some PRs are not affected by the changes and can be already reviewed?)

dhruv commented:
@theStack awesome! I'm working on changes you mentioned. In the meantime #23432 and upstream elligator squared changes in secp are unaffected and you can start review there. #23561 will have very minimal changes so you can review now and the updates will be easy to review. #20962 and #23233 will be radically different.

laanwj commented:
> or is it better to just wait for the final spec / updated BIP

Waiting is not a good option :smile: I think it's unavoidable that the both the BIP and implementation will still evolve a bit during review and testing, as new issues are discovered.

theStack commented:
> > or is it better to just wait for the final spec / updated BIP
> 
> Waiting is not a good option 😄 I think it's unavoidable that the both the BIP and implementation will still evolve a bit during review and testing, as new issues are discovered. In FOSS there's pretty much never a straightforward spec → implementation flow.

Agree! My concern was that the undergoing spec changes are possibly so intense that the current publicly available BIP barely has any resemblance with the ~~final~~ updated version anymore and reviewing/testing current PRs would not be a well-invested time. But fortunately, that doesn't seem to be the case (https://github.com/bitcoin/bitcoin/pull/24545#issuecomment-1141408247, https://bip324.com/sections/code-review/), i.e. no excuses for waiting :)

dhruv committed:
RFC8439 implementation and tests

dhruv committed:
Adding forward secure FSChaCha20

dhruv committed:
Enable ECDH computation on secp256k1 keys

dhruv committed:
Bench test for ECDH

ExperiBass commented:
> From what I can tell, my node fails to connect to another p2p v2 peer (likely @laanwj's, we haven't fully confirmed that yet) and stops accepting inbound messages from any other peer cutting it off from the network entirely.

I also seem to have this issue. Relevant logs:

```
2022-07-22T13:55:24Z received: inv (109 bytes) peer=19
2022-07-22T13:55:24Z got inv: tx cae283722e02ae9aa0e0da00b8064b76bf05583d17d8b8b4bc8dc84578a77a3b  new peer=19
2022-07-22T13:55:24Z got inv: tx ee58f2752d52e7fc79e9e19ddc9b592c4fb105cced7f5c245175eb0615bf2139  new peer=19
2022-07-22T13:55:24Z got inv: tx 3f1a7195723b7a711aac44c43ff12a6a3e360bacf52f685d0c5f1e706c64de3f  new peer=19
2022-07-22T13:55:24Z socket closed for peer=16
2022-07-22T13:55:24Z disconnecting peer=16
2022-07-22T13:55:24Z Cleared nodestate for peer=16
2022-07-22T13:55:25Z received: version (102 bytes) peer=23
2022-07-22T13:55:25Z sending wtxidrelay (0 bytes) peer=23
2022-07-22T13:55:25Z sending sendaddrv2 (0 bytes) peer=23
2022-07-22T13:55:25Z sending verack (0 bytes) peer=23
(no more inbound)
```
[Full log](https://github.com/bitcoin/bitcoin/files/9169486/324-disconnect.txt)

dhruv committed:
Fuzz test for ECDH

dergoegge reviewed (commented):
I think you can drop the `net_processing` changes if you don't call `InitializeNode` for outbound connections until it is clear which transport version will be used (see my review comments, they only make sense together). *I think*  this is safe w.r.t only calling `InitializeNode` once but please double check.

dhruv committed:
p2p: BIP324 shapable key exchange

dhruv committed:
Allow for RFC8439 AD in cipher suite interface

dhruv committed:
Expose BIP324CipherSuite AAD via transport classes

dhruv committed:
test: Functional test for opportunistic encryption

dhruv commented:
Changes in this push:
- Bring in upstream changes
- Added shapable handshake
- Added a functional test

dhruv commented:
@ExperiBass, I have worked with @0xB10C and @stratospher to find and fix the thread locking issue. Are you able to try again with the latest code?

This push:
- Fixed thread locking issue that happened when the Initiator tried to downgrade a v2 BIP324 connection to a v1 plaintext connection whilst the outbound slots were full.
- Fixed another bug where the Initiator would bot send the VERSION message on a v2 connection if there were no more inbound bytes to process. This wasn't caught by the functional test because there were more inbound bytes to process.

Next up, comments from @dergoegge.

ExperiBass commented:
> @ExperiBass, I have worked with @0xB10C and @stratospher to find and fix the thread locking issue. Are you able to try again with the latest code?

So far so good, i havent seen any crashes for the past 24 hours. Looks like its fine 👍🏾

dhruv commented:
Latest push:
- Rebased
- Nomenclature update
- Bring in upstream changes
- Interpret encrypted length as length of contents rather than header + contents
- Rekey salt length is now 23 bytes instead of 32 bytes
- New test vectors, including a trivial one-byte vector and a large vector
- Authentication of garbage sent by initiator in the shapable handshake

dhruv committed:
test: BIP324 test vectors

dhruv committed:
rpc: Expose transport type via getpeerinfo

dhruv committed:
rpc: Expose BIP324 session id via getpeerinfo

dhruv commented:
Latest push brings all the code up to date with the [released BIP draft](https://github.com/bitcoin/bips/pull/1378). Changes:

#### Handshake shapability

- Bidirectional garbage and terminators
- Early key response by responder upon first mismatching byte
- Allowing for multiple rounds in key exchange
- Remove bias away from NETWORK_MAGIC || "version\x00" for ElligatorSwift encodings
- Remove garbage terminator from the garbage authentication packet as it is implicitly authenticated

#### Bugfixes

- An already established v1 connection peer should not send out an ellswift key if a message has a bad v1 header
- The previous changes to test_framewotk.py made all tests fail with a small probability, that's now fixed

#### Test vectors
- One-sided test vectors consistent with the BIP (the BIP actually needs updating which is WIP)
- Test vectors in a json file instead of making net_tests.cpp even longer

#### RPC

- Expose transport protocol type via `getpeerinfo`
- Expose v2 session id via `getpeerinfo`

Also rebased the PR.

Ready for further review. Changes in this push are not backwards compatible with the previous code(handshake shapability changes) so it'd be nice if everyone running the branch could run the latest version.

dhruv commented:
Pushed again: tests against previous releases (which don't run on my personal fork for some reason) were failing because of the new optional argument to rpc:`addnode` not existing in older clients.

dhruv commented:
Latest push adds a downgrade to v1 upon inactivity within the peer timeout period. Also changed the order of commits.

`git diff e823a19d59 062e51c2f0`

jamesob commented:
Docker image ([`jamesob/bitcoind:bip324-enable`](https://hub.docker.com/layers/jamesob/bitcoind/bip324-enable/images/sha256-89cbe61ab22aa55424624ca2ceaa1f019e8757f8f55fdfef56a77eb4261a3069?context=explore)) built for this branch as of 062e51c2f051dc68197bc21535fd06b3e01bdc21, in case that helps anyone test.

rot13maxi commented:
built the `dhruv:bip324-enable` at commit `062e51c2f051dc68197bc21535fd06b3e01bdc21` on an m1 mac. No problems building. Started a new node with `-v2transport=1` turned on. At different points of IBD, I added/removed peers to make sure I could get blocks from all v1 peers, v1 + v2 peers, v2-only peers. Once the node was synced, connected to only v2 peers and tested broadcasting transactions. Also compared mempool contents and chaintips with a v23 node running the v1 transport. So far, everything works as expected.

jsarenik commented:
Thanks for doing this! I compiled it myself with one little [unrelated patch](https://github.com/jsarenik/bitcoin/tree/jsn/bip324-enable) on Alpine Linux (musl-libc) x86_64. It is running at [be.anyone.eu.org](https://be.anyone.eu.org/), all three Bitcoin networks ­— mainnet, testnet and signet (and also [Plebnet Playground](https://plebnet.wiki/wiki/Plebnet_Playground), which is a custom signet). IPv4, Tor, I2P. Feel free to test with my node running on residential 30 Mbps DSL connection 24/7.

dhruv commented:
Latest push:
- Bringing in upstream optimizations to https://github.com/bitcoin-core/secp256k1/pull/1129

dhruv commented:
Latest push:
- Bringing in upstream rebase to #26153 
- Fixed a bug where a disconnection after receiving some bytes but prior to the end of the key exchange phase would result in a downgrade when it should not

dergoegge reviewed (changes requested):
Concept ACK - Thanks to everyone working on this!

My review feedback will be limited to the networking changes as I can't really judge the soundness of the underlying crypto used (Cryptography 1 by Dan Boneh didn't quite get me there yet :D).

The current approach here feels a little "slapped on" to me. I am surprised by all the special casing around handling v2 connections which I think mostly stems from the fact that currently v1 and v2 connections are being accepted on the same port. Is that really required/desirable for some reason?

I kind of expected to see a similar approach to how we handle connections from different networks (e.g. Tor, CJDNS, I2P). That is, we listen for v2 connection on a dedicated port in a separate thread (or a separate process but that could also be done later or never) that handles all the initial v2 connection setup (i.e. key exchange). Once a v2 connection has been successfully established the thread registers the new connection with the connection manager. This would lead to the changes here being much more self contained and remove the need for the special casing around v2 connection setup.

jsarenik commented:
Updated my nodes to 8f7714e33804b6cf96fedbe05751d1a700e1e8d1

Node1's addresses (no Tor anymore) are visible via [node-details](https://be.anyone.eu.org/node-details).

Node2's addresses are:
  - Tor: `3xept3fmslplcm4hu4gnbpquz3d5twqg2m5hqeimbvuzykiuxan5kgad.onion`
  - Yggdrasil (which thinks it is IPv6): `200:40c3:3ffe:680:a5b2:9d41:3928:fb73`

jsarenik commented:
> My review feedback will be limited to the networking changes as I can't really judge the soundness of the underlying crypto used (Cryptography 1 by Dan Boneh didn't quite get me there yet :D).

Much appreciated. I did not even finish that first course yet :)

> The current approach here feels a little "slapped on" to me. I am surprised by all the special casing around handling v2 connections which I think mostly stems from the fact that currently v1 and v2 connections are being accepted on the same port. Is that really required/desirable for some reason?

Practically speaking for myself, it actually only makes sense when v1 and v2 live on the same port. Example: a tunelled IPv6 connection may be doing some L7 blocking, but when the filter sees other full-duplex communication which it can not decipher, it may be less willing to block the whole port since there is other noise on it.

> I kind of expected to see a similar approach to how we handle connections from different networks (e.g. Tor, CJDNS, I2P). ...

Tor, CJDNS and I2P are both layers and addressing schemes at the same time. CJDNS (and [Yggdrasil](https://github.com/bitcoin/bitcoin/pull/23531)) are merely subsets of IPv6 address space for addressing, but they do their own thing encrypting the contents of the packets.

V2 IMHO makes sense on clearnet IPv4 or IPv6. Even if the other transports go "cleartext" on the user-level, still no one can see anything during the transfer and V2 will not make it any better, just use more CPU for nothing (double-encryption, though useful for testing by being able to add an onion address as a V2 peer).

dhruv commented:
Latest push:
- Rebased upstream changes
- New rekey ratchet mechanism (forward security) based on ChaCha20 instead of SHA256. BIP changes are WIP and coming soon.

Note for users running mainnet nodes for testing: **This is a breaking change in the protocol. So you will need to update code to peer with nodes that have updates.**

rot13maxi commented:
@dhruv can you elaborate on why the move to chacha20?

dhruv commented:
@rot13maxi It's a minor improvement: ChaCha20 is cheaper than SHA256 and we are already using it. For the length encryption, it can be done without even initializing a new instance. The payload encryption is also already using RFC8439 ChaCha20-Poly1305.

dhruv commented on src/net.h:
```diff
@@ -526,7 +653,9 @@ class CNode
-          CNodeOptions&& node_opts = {});
+          CNodeOptions&& node_opts = {},
+          bool prefer_p2p_v2 = false,
+          TransportProtocolType transport_type_in = TransportProtocolType::V1);
```
Done.

dhruv commented on src/net.h:
```diff
@@ -592,13 +721,21 @@ class CNode
+    std::array<std::byte, BIP324_GARBAGE_TERMINATOR_LEN> v2_sent_garbage_terminator;
+    std::array<std::byte, BIP324_GARBAGE_TERMINATOR_LEN> v2_recv_garbage_terminator;
+    std::vector<std::byte> v2_garbage_bytes_recd;
+    bool v2_keys_derived{false};
```
Done. I chose to break it down into two different structs -- one for private members and one for public ones.

dhruv commented:
Latest push:
- Rebased
- Addressed [review comments](https://github.com/bitcoin/bitcoin/pull/24545#pullrequestreview-1194881025)

dhruv commented:
Latest push:
- Rebased changes from #23432

dhruv commented:
Rebased changes from #26153

dhruv commented:
Rebased. Brought in upstream changes.

dhruv commented:
Rebased.

vostrnad commented:
Is there a reason the `v2transport` option should be off by default? Does it even need to exist at all? I was accidentally running this for a few days with v2 turned off and was confused why nodes advertising `P2P_V2` weren't actually using the v2 protocol. Everything works fine now.

Also, it would be nice to have the `transport_protocol_type` and `v2_session_id` fields displayed in the GUI, although that might possibly belong to a separate PR in the GUI repo (not sure what the development process dictates here).

dhruv commented:
Rebased with master. Brought in changes to #23233, #23561 and #23432. **The changes to the BIP [here](https://github.com/bitcoin/bips/pull/1428) are implemented and result in a breaking change for everyone running v2 nodes on mainnet.**

vostrnad commented:
After updating, node is stuck in a loop trying to connect to manually added nodes that haven't been updated yet, without downgrading to a v1 connection. It might be just fine, but I thought I'd mention it.

dhruv commented:
@vostrnad this is expected because the breaking change is in the p2p layer and not the transport. So a v2 connection is successfully established, and likely the inbound peer (not updated) never sees something that it interprets as the VERSION message. It should have timed out/failed eventually though. Did it do that? This isn't currently considered a downgrade scenario because a v2 transport connection was established. If we downgrade to v1 on p2p unresponsiveness, my thinking is that would lead to downgrades in scenarios where a peer is trying to evict, etc.

dhruv committed:
Merge branch 'bip324-cipher-suite' into bip324-net-v2

dhruv committed:
Merge commit '963f9a5c6159d985eb16b115a7d27027074827ed' into bip324-ellsq

dhruv committed:
Squashed 'src/secp256k1/' changes from bdf39000b9..8034c67a48

8034c67a48 Add doc/ellswift.md with ElligatorSwift explanation
e90aa4e62e Add ellswift testing to CI
131faedd8a Add ElligatorSwift ctime tests
198a04c058 Add tests for ElligatorSwift
9984bfe476 Add ElligatorSwift benchmarks
f053da3ab7 Add ellswift module implementing ElligatorSwift
76c64be237 Add functions to test if X coordinate is valid
aff948fca2 Add benchmark for key generation
5ed9314d6d Add exhaustive tests for ecmult_const_xonly
b69fe88d5e Add x-only ecmult_const version for x=n/d
427bc3cdcf Merge bitcoin-core/secp256k1#1236: Update comment for secp256k1_modinv32_inv256
647f0a5cb1 Update comment for secp256k1_modinv32_inv256
5658209459 Merge bitcoin-core/secp256k1#1228: release cleanup: bump version after 0.3.0
28e63f7ea7 release cleanup: bump version after 0.3.0

git-subtree-dir: src/secp256k1
git-subtree-split: 8034c67a48dc1334bc74ee4ba239111a23d9789e

dhruv committed:
Merge branch 'bip324-cipher-suite' into bip324-handshake

dhruv committed:
Merge commit '984b3d0a1faabc1657e0dd33432fdadfd78335a9' into bip324-handshake

dhruv committed:
Squashed 'src/secp256k1/' changes from bdf39000b9..8034c67a48

8034c67a48 Add doc/ellswift.md with ElligatorSwift explanation
e90aa4e62e Add ellswift testing to CI
131faedd8a Add ElligatorSwift ctime tests
198a04c058 Add tests for ElligatorSwift
9984bfe476 Add ElligatorSwift benchmarks
f053da3ab7 Add ellswift module implementing ElligatorSwift
76c64be237 Add functions to test if X coordinate is valid
aff948fca2 Add benchmark for key generation
5ed9314d6d Add exhaustive tests for ecmult_const_xonly
b69fe88d5e Add x-only ecmult_const version for x=n/d
427bc3cdcf Merge bitcoin-core/secp256k1#1236: Update comment for secp256k1_modinv32_inv256
647f0a5cb1 Update comment for secp256k1_modinv32_inv256
5658209459 Merge bitcoin-core/secp256k1#1228: release cleanup: bump version after 0.3.0
28e63f7ea7 release cleanup: bump version after 0.3.0

git-subtree-dir: src/secp256k1
git-subtree-split: 8034c67a48dc1334bc74ee4ba239111a23d9789e

dhruv committed:
Merge branch 'bip324-handshake' into bip324-enable

dhruv committed:
Merge branch 'bip324-ellsq' into bip324-enable

dhruv commented:
Rebased.

fanquake commented:
Closing for now. This will be picked up again later. BIP324 review attention should be directed towards https://github.com/bitcoin/bitcoin/pull/27479 and https://github.com/bitcoin-core/secp256k1/pull/1129.
